### ISSUE TITLE
Which Loss?

### ISSUE THREAD

**Issue by sockfish** [UKPLab/sentence-transformers#792]
Say I have a ~100 datasets, each of them containing ~10000 sentences, classified into ~100 classes. The classes are different for each dataset so sentences across datasets cannot be compared.

Say I want to fine-tune using this data. Which loss would you try first?

I guess it would be one of the TripletLoss ones? Or MegaBatchMarginLoss?

Thanks!

**Comment by nreimers** [comment]
I would merge the datasets and make the labels distinct for the different datasets, e.g. dataset1 has the labels 1-100, dataset2 the labels 101-200 etc.

Then I would use BatchHard losses with large batch sizes. Note, you need to ensure that every label in a batch has at least two samples from the same label. You can ensure this with a custom dataset class:
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/other/training_batch_hard_trec.py

**Comment by sockfish** [comment]
Great, thanks! There is one thing that concerns me though: Label A in dataset 1 may be very similar to label B in dataset 2. For example: dataset 1 has label A:"Awesome movie" and dataset 2 has label B:"Great movie". 

So given an anchor in A, the most dissimilar sentence within A may very well be more different to the anchor than the most similar sentence from label B. The only thing I know with some certainty is that the labels within one dataset should be different from each other.

So I guess I should make sure that batches are always picked from within one dataset

**Comment by nreimers** [comment]
Yes, that would be an issue.

In that case, you can try multi-task learning with your different datasets:
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/other/training_multi-task.py


An issue here is that samples from dataset 1 are never compared with samples from dataset 2. It could happen that the vector spaces overlap. E.g. label A from dataset 1 is in the lower left corner. In that corner, there is also label B from dataset 2. However, label A and B are quite different. But as samples from these two are never in the same batch, it will not be learned during training
